ocaml/xapi/xapi_pgpu_helpers.ml

```diff
@@ -50,20 +50,33 @@ let get_allocated_VGPUs ~__context ~self =
 
 let assert_VGPU_type_allowed ~__context ~self ~vgpu_type =
   assert_VGPU_type_enabled ~__context ~self ~vgpu_type;
-  (match get_allocated_VGPUs ~__context ~self with
-   | [] -> ()
-   | resident_VGPU :: _ ->
-     let running_type =
-       Db.VGPU.get_type ~__context ~self:resident_VGPU
-     in
-     if running_type <> vgpu_type
-     then raise (Api_errors.Server_error (
-         Api_errors.vgpu_type_not_compatible_with_running_type,
-         [
-           Ref.string_of self;
-           Ref.string_of vgpu_type;
-           Ref.string_of running_type;
-         ])))
+  let allocated_vgpu_list = get_allocated_VGPUs ~__context ~self in
+  (** Now check whether the requested type is permitted *)
+  match allocated_vgpu_list with
+  | [] -> () (* Not allocated on this pgpu, does not need to check compatibility*)
+  | hd::tail ->
+    let get_compatible_types_on_pgpu vgpu=
+      let vgpu_type = Db.VGPU.get_type ~__context ~self:vgpu in
+      Db.VGPU_type.get_compatible_types_on_pgpu ~__context ~self:vgpu_type in
+    (* The new vGPU type must in the compatible_types_on_pgpu of all the allocated vGPUs*)
+    let grant_vgpu_type_list = List.fold_left
```

Comment:
This code is paying a price for using lists instead of sets. Why do we need to `sort_uniq` before we can pass the result to `List.intersect`?